pytensor/gradient.py

```diff
@@ -2052,6 +2052,72 @@ def hessian(cost, wrt, consider_constant=None, disconnected_inputs="raise"):
     return as_list_or_tuple(using_list, using_tuple, hessians)
 
 
+def hessian_vector_product(cost, wrt, p, **grad_kwargs):
+    """Return the expression of the Hessian times a vector p.
+
+    Parameters
+    ----------
+    cost: Scalar (0-dimensional) variable.
+    wrt: Vector (1-dimensional tensor) 'Variable' or list of Vectors
+    p: Vector (1-dimensional tensor) 'Variable' or list of Vectors
+        Each vector will be used for the hessp wirt to exach input variable
+    **grad_kwargs:
+        Keyword arguments passed to `grad` function.
+
+    Returns
+    -------
+    :class:` Vector or list of Vectors
+        The Hessian times p of the `cost` with respect to (elements of) `wrt`.
+
+    Examples
+    --------
+
+    .. testcode::
+
+        import numpy as np
+        from scipy.optimize import minimize
+
+        from pytensor import function
+        from pytensor.tensor import vector
+        from pytensor.gradient import jacobian, hessian_vector_product
+
+        x = vector('x')
+        p = vector('p')
+
+        rosen = (100 * (x[1:] - x[:-1] ** 2) ** 2 + (1 - x[:-1]) ** 2).sum()
+        rosen_hessp = hessian_vector_product(rosen, x, p)
+        rosen_jac = jacobian(rosen, x)
```

Comment:
Why should we call jacobian here? `grad` should be enough as the function has a scalar value.

Reply:
They're the same, I didn't choose jacobian on purpose

Reply:
I find this more readable for this example as it goes into the jac argument in the scipy minimize. Strong objection?

Reply:
If jacobian doesn't introduce a scan for a scalar argument than I think it is ok, but I think it still is useful to keep users away from calling jacobian unnecessarily, because usually it is so much more expensive. But I don't care all that much :-)

Reply:
for ndim=0 it calls grad